ui/src/features/annotator/tools/utils.ts

```diff
@@ -173,3 +172,173 @@ export const getClampedBoundingBox = (point: Point, boundingBox: RegionOfInteres
+        case 'rect':
+            return {
+                shapeType: 'rect',
+                x: shape.x,
+                y: shape.y,
+                width: shape.width,
+                height: shape.height,
+            };
+        default:
+            throw new Error('Unknown shape type');
+    }
+}
+
+const removeOffPointsRect = (rect: Rect, roi: RegionOfInterest): Rect => {
+    const { x, y, width, height } = roi;
+
+    let newRect: Rect = {
+        ...rect,
+    };
+
+    if (rect.x < x) {
+        newRect = {
+            ...newRect,
+            x,
+            width: newRect.width - (x - rect.x),
+        };
+    }
+
+    if (rect.x + rect.width > x + width) {
+        const diff = rect.x + rect.width - x - width;
+
+        newRect = {
+            ...newRect,
+            width: rect.width - diff,
+        };
+    }
+
+    if (rect.y < y) {
+        newRect = {
+            ...newRect,
+            y,
+            height: rect.height - (y - rect.y),
+        };
+    }
+
+    if (newRect.y + rect.height > y + height) {
+        const diff = rect.y + rect.height - y - height;
+
+        newRect = {
+            ...newRect,
+            height: rect.height - diff,
+        };
+    }
+
+    return newRect;
+};
+
+const removeOffLimitPointsPolygon = (shape: Shape, roi: RegionOfInterest): Polygon => {
+    const { width, height, x, y } = roi;
+    const getRect = (rx: number, ry: number, rWidth: number, rHeight: number): Rect => ({
+        x: rx,
+        y: ry,
+        width: rWidth,
+        height: rHeight,
+        shapeType: 'rect',
+    });
+    // `eraserSize` Builds and positions rect shapes around ROI limits (top, left, right, bottom),
+    // finally `getShapesDifference` will use those rects to calc and remove offline polygons
+    const eraserSize = 10;
+    const topRect = getRect(x - eraserSize, y - eraserSize, width + eraserSize * 3, eraserSize);
+    const leftRect = getRect(x - eraserSize, y - eraserSize, eraserSize, height * 2);
+    const rightRect = getRect(x + width, y - eraserSize, eraserSize, height * 2);
+    const bottomRect = getRect(x - eraserSize, y + height, width + eraserSize * 3, eraserSize);
+
+    return [leftRect, bottomRect, rightRect, topRect].reduce(
+        (accum, current) => getShapesDifference(roi, accum, current),
+        shape
+    ) as Polygon;
+};
+
+const calculateRectanglePoints = (shape: BoundingBox): ClipperPoint[] => {
+    const { x: X, y: Y, width, height } = shape;
+    const topLeftPoint = { X, Y };
+    const topRightPoint = { X: X + width, Y };
+    const bottomLeftPoint = { X, Y: Y + height };
+    const bottomRightPoint = { X: X + width, Y: Y + height };
+
+    return [topLeftPoint, topRightPoint, bottomRightPoint, bottomLeftPoint];
+};
+
+const convertPolygonPoints = (shape: Polygon): ClipperPoint[] => {
+    return shape.points.map(({ x, y }: Point) => ({ X: x, Y: y }));
+};
+
+const transformToClipperShape = (shape: Shape): ClipperShape => {
+    switch (true) {
```

Comment:
I’m pretty sure I wrote this, but it might be simpler to just go with a regular `if/else`